### PR TITLE
Added counting_sort.cpp which implements Counting Sort for integers for DWOC 2020-21

### PR DESCRIPTION
**Pull request by Sarthak-9**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

### Description

Contributed to open source

### Checklist

- [ #] I've read the contribution guidelines.
- [ #] I've checked the issue list before deciding what to submit.
- [ #] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

Implemented Count Sort #1776

### CODE REVIEW

File: C-Plus-Plus/sort/counting_sort.cpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+// where n is the number of elements in input array and k is the range of input.
+//
+
+
+#include <iostream>
+
+using namespace std;
+
+int Max(long long int A[], long long int n){
+    long long int max = -32768;
```
Comment (atarax665) [inline C-Plus-Plus/sort/counting_sort.cpp]: Its better to assign the 1st element of the array (A[0]) as max initially while checking for maximum element in an array.

File: C-Plus-Plus/sort/counting_sort.cpp [hunk 2]
```diff
@@ -0,0 +1,72 @@
+//
```
Comment (atarax665) [inline C-Plus-Plus/sort/counting_sort.cpp]: Use multi-line comments instead of single line comments.

File: C-Plus-Plus/sort/counting_sort.cpp [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        }
+    }
+    return max;
+}
+
+void CountSort(long long int A[], long long int n){
+    long long int max = Max(A, n);
+    int* count = new int [max + 1];
+
+    for (long long int i=0; i<max+1; i++){
```
Comment (atarax665) [inline C-Plus-Plus/sort/counting_sort.cpp]: Add comment before each  loop to explain what it does in the code.

File: C-Plus-Plus/sort/counting_sort.cpp [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+    CountSort(A, n);
+
+    cout<<"The sorted array is :"<<endl;
+    for(long long int i=0;i<n;i++){
+        cout<<A[i]<<" ";
+    }
+    cout<<endl;
+    return 0;
+}
```
Comment (atarax665) [inline C-Plus-Plus/sort/counting_sort.cpp]: Add Sample Input/Output in commented form, refer to any recent merged C++ PR for reference.

**Review by atarax665 — CHANGES_REQUESTED** [review]
Format your code according to [C++ style guide](https://google.github.io/styleguide/cppguide.html).

File: C-Plus-Plus/README.md [hunk 5]
```diff
@@ -118,7 +118,7 @@
 - [Shell Sort](sort/ShellSort.cpp)
 - [Topological Sort](sort/Topological_Sorting_DFS.cpp)
 - [Radix Sort](sort/radix_sort.cpp)
-
+- [Counting Sort](sort/counting_sort.cpp)
```
Comment (atarax665) [inline C-Plus-Plus/README.md]: Add a blank line after this.

**Review by atarax665 — CHANGES_REQUESTED** [review]

**Comment by Sarthak-9** [comment]
I'll resolve these soon.

**Review by atarax665 — CHANGES_REQUESTED** [review]
Format the code according to [C++ style guide](https://google.github.io/styleguide/cppguide.html).

**Comment by Sarthak-9** [comment]
> Format the code according to [C++ style guide](https://google.github.io/styleguide/cppguide.html).

Could you please elaborate a bit. I have made some changes but still not sure.

**Review by iamrajiv — CHANGES_REQUESTED** [review]
@Sarthak-9 Your code is not formatted properly. Format like this

```c++
/*
 Created by Sarthak-9 on 30-12-2020.
 Count Sort Algorithm for Sorting an array of small integers quickly.

 Worst complexity: O(n+k)
 Average complexity: O(n+k)
 Space complexity: O(n+k)
 where n is the number of elements in input array and k is the range of input.
*/

#include <iostream>

using namespace std;

int Max(long long int A[], long long int n) {
	long long int max = A[0];

	// finds the maximum element of the array
	for (long long int i = 0; i < n; i++) {
		if (A[i] > max) {
			max = A[i];
		}
	}
	return max;
}

void CountSort(long long int A[], long long int n) {
	long long int max = Max(A, n);
	int* count = new int [max + 1];

	// initializes the count array with all elements to zero
	for (long long int i = 0; i < max + 1; i++) {
		count[i] = 0;
	}

	// updates the count array with frequencies of given numbers
	for (long long int i = 0; i < n; i++) {
		count[A[i]]++;
	}

	long long int i = 0;
	long long int j = 0;

	// updates A with sorted elements
	while (j < max + 1) {
		if (count[j] > 0) {
			A[i++] = j;
			count[j]--;
		} else {
			j++;
		}
	}

	delete [] count;
}

int main() {

	long long int n;
	long long int A[100000];
	cout << "Enter the size of array" << endl;
	cin >> n;
	cout << "Enter the elements" << endl;

	// inputs numbers from the user
	for (long long int i = 0; i < n; i++) {
		cin >> A[i];
	}

	CountSort(A, n);

	cout << "The sorted array is :" << endl;

	// prints the sorted array
	for (long long int i = 0; i < n; i++) {
		cout << A[i] << " ";
	}
	cout << endl;
	return 0;
}

/*
  Input:
  Enter the size of array
  10
  Enter the elements
  45 86 27 38 59 12 86 72 125 38

  Output:
  The sorted array is :
  12 27 38 38 45 59 72 86 86 125
*/
```

**Comment by Sarthak-9** [comment]
Please review

**Review by atarax665 — CHANGES_REQUESTED** [review]
Your current implementation won't work for negative integers, please modify your code for standard array input.

**Comment by Sarthak-9** [comment]
I optimized the solution for negative numbers. Please review .

**Comment by Sarthak-9** [comment]
@atarax665  @shraddhavp Please review

**Comment by Sarthak-9** [comment]
> @atarax665 @shraddhavp Please review

Any changes?

**Review by iamrajiv — APPROVED** [review]

**Review by atarax665 — APPROVED** [review]
Resolve the conflict ASAP so that I can merge this PR.

**Comment by Sarthak-9** [comment]
@atarax665 Conflicts resolved

**Pull request merged**